### PR TITLE
Add development and contributing to docs

### LINKED ISSUE #118: Add contributing and developing docs
Topics to cover

- contributing guide (even if only approximate for now)
- testing philosophy/style (or lack thereof)
- resources for those who want to learn more (e.g. about git, mocking)

### PR DESCRIPTION
**Pull request by znicholls**
- [x] Closes #118 
 - [x] Fully documented, including `CHANGELOG.rst`

### CODE REVIEW

File: docs/source/development.rst [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+Regression tests are there to ensure that we know when outputs will change (sometimes they should change, we just want to make sure that this change is deliberate not accidental).
+Regression tests don't require too much code generally, but they may run lots of the code base and hence take a little while to run.
+
+(We are in the process of making the distinction between regression and integration tests clearer, see `#120 <https://github.com/MESMER-group/mesmer/issues/120>`_).
+
+
+Formatting
+----------
+
+To help us focus on what the code does, not how it looks, we use a couple of automatic formatting tools.
```
Comment (mathause) [inline docs/source/development.rst:121]:
```suggestion
To help us focus on what the code does, not how it looks, we use a couple of automatic formatting tools.
We use the following tools:
- `isort <https://github.com/PyCQA/isort>`_ to sort import statements
- `black <https://github.com/psf/black>`_ to auto-format the code
- `flake8 <https://flake8.pycqa.org/en/latest/>`_ to check the format and small errors

```

File: Makefile [hunk 2]
```diff
@@ -35,6 +35,11 @@ export PRINT_HELP_PYSCRIPT
 help:
 	@python -c "$$PRINT_HELP_PYSCRIPT" < $(MAKEFILE_LIST)
 
+.PHONY: format
+format: $(VENV_DIR)  ## auto-format the code using relevant tools
+	make isort
+	make black
```
Comment (mathause) [inline Makefile:41]:
```suggestion
	make black
	make flake8
```
I always do all three...

File: docs/source/development.rst [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+As a contributor, please follow a couple of conventions:
+
+- Create issues in the `MEMSER issue tracker`_ for changes and enhancements, this ensures that everyone in the community has a chance to comment
+- Be welcoming to newcomers and encourage diverse new contributors from all backgrounds: see the `Python Community Code of Conduct <https://www.python.org/psf/codeofconduct/>`_
+- Only push to your own branches, this allows people to force push to their own branches as they need without fear or causing others headaches
+- Start all pull requests as draft pull requests and only mark them as ready for review once they've been rebased onto master, this makes it much simpler for reviewers
+- Try and make lots of small pull requests, this makes it easier for reviewers and faster for everyone as review time grows exponentially with the number of lines in a pull request
+
+Getting setup
+-------------
```
Comment (mathause) [inline docs/source/development.rst:35]:
```suggestion
Development setup
-----------------
```

File: docs/source/development.rst [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    #. Create a conda environment to use with MESMER ``conda create --name mesmer``
+    #. Activate your conda environment ``conda activate mesmer``
+    #. Install the development dependencies (very important, make sure your conda environment is active before doing this) ``conda install -y --file environment.yml && pip install --upgrade pip wheel && pip install -e .[dev]``
+
+#. Make sure the tests pass by running ``make test``, if that fails the commands are
+
+    #. Activate your conda environment ``conda activate mesmer``
+    #. Run the unit and integration tests ``pytest --cov -r a --cov-report term-missing``
+
+Getting help
```
Comment (mathause) [inline docs/source/development.rst:52]: Is this for users or for developers?

**Review by mathause — COMMENTED** [review]
Looks good - thanks. Maybe the `Development tools` section should be before `Development setup`?

File: docs/source/development.rst [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+Please report issues or discuss feature requests in the `MEMSER issue tracker`_.
+If your issue is a feature request or a bug, please use the templates available, otherwise, simply open a normal issue :)
+
+As a contributor, please follow a couple of conventions:
+
+- Create issues in the `MEMSER issue tracker`_ for changes and enhancements, this ensures that everyone in the community has a chance to comment
+- Be welcoming to newcomers and encourage diverse new contributors from all backgrounds: see the `Python Community Code of Conduct <https://www.python.org/psf/codeofconduct/>`_
+- Only push to your own branches, this allows people to force push to their own branches as they need without fear or causing others headaches
+- Start all pull requests as draft pull requests and only mark them as ready for review once they've been rebased onto master, this makes it much simpler for reviewers
```
[the remark this answers is not in the document]
Comment (yquilcaille) [reply docs/source/development.rst:31]: Typo:

> `ME**SM**ER issue tracker`

File: docs/source/development.rst [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+- tutorials (or support questions which, once solved, result in a new tutorial :D)
+- blog posts
+- improving the documentation
+- bug reports
+- feature requests
+- pull requests
+
+Please report issues or discuss feature requests in the `MEMSER issue tracker`_.
+If your issue is a feature request or a bug, please use the templates available, otherwise, simply open a normal issue :)
```
[the remark this answers is not in the document]
Comment (znicholls) [reply docs/source/development.rst:24]: Good point, will remove

File: docs/source/development.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (znicholls) [reply docs/source/development.rst:31]: Will fix

File: docs/source/development.rst [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+To get setup as a developer, we recommend the following steps (if any of these tools are unfamiliar, please see the resources we recommend in `Development tools`_):
+
+#. Install conda and make
+#. Run ``make conda-environment``, if that fails you can try doing it manually
+
+    #. Change your current directory to MESMER's root directory (i.e. the one which contains ``README.rst``), ``cd mesmer``
+    #. Create a conda environment to use with MESMER ``conda create --name mesmer``
+    #. Activate your conda environment ``conda activate mesmer``
+    #. Install the development dependencies (very important, make sure your conda environment is active before doing this) ``conda install -y --file environment.yml && pip install --upgrade pip wheel && pip install -e .[dev]``
```
[the remark this answers is not in the document]
Comment (znicholls) [reply docs/source/development.rst:45]: `pip install --upgrade pip wheel` is legacy (could be changed to conda)

`pip install -e .[dev]` is it possible to do editable installs with conda? I've always just done pip as it's simple and I know it...

File: docs/source/development.rst (hunk 4 above)
Comment (znicholls) [reply docs/source/development.rst:52]: For developers, user help goes elsewhere

File: docs/source/development.rst [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+- `Git <http://swcarpentry.github.io/git-novice/>`_
+
+- `Make <https://swcarpentry.github.io/make-novice/>`_
+
+- `Conda virtual environments <https://medium.freecodecamp.org/why-you-need-python-environments-and-how-to-manage-them-with-conda-85f155f4353c>`_
+
+- `Tests <https://semaphoreci.com/community/tutorials/testing-python-applications-with-pytest>`_
+
+    - we use a blend of `pytest <https://docs.pytest.org/en/latest/>`_ and the inbuilt Python testing capabilities for our tests so checkout what we've already done in ``tests`` to get a feel for how it works
```
[the remark this answers is not in the document]
Comment (znicholls) [reply docs/source/development.rst:75]: Yep

File: docs/source/development.rst [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+- `Continuous integration (CI) <https://docs.travis-ci.com/user/for-beginners/>`_
+
+    - we use `Travis CI <https://travis-ci.com/>`_ for our CI but there are a number of good providers
```
[the remark this answers is not in the document]
Comment (znicholls) [reply docs/source/development.rst:79]: Thanks, will update

**Comment by znicholls** [comment]
> Maybe the `Development tools` section should be before `Development setup`?

I thought about that but I don't think it flows as nicely

**Comment by mathause** [comment]
Something goes wrong importing xarray on rtd. I half-hartedly tried to create the env locally but could not reproduce the problem. These are the most annoying issues.

**Comment by mathause** [comment]
The docs build picked up python 3.10 - it's too early for this - see  #129. (What is strange, however, the problem only occurred when calling `make html` while on the command line it worked just fine).

File: docs/source/development.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (mathause) [reply docs/source/development.rst:44]: Yes you can pip-install stuff in a conda environment. I try restrict this to the minimum, so conda has the best chance to solve/update the env.

[the remark this answers is not in the document]
Comment (znicholls) [reply docs/source/development.rst:44]: Yes me too but there are some things that aren't on conda etc. so I've always ended up having to go with this hybrid approach, as painful as it is

File: docs/source/development.rst [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+To get setup as a developer, we recommend the following steps (if any of these tools are unfamiliar, please see the resources we recommend in `Development tools`_):
+
+#. Install conda and the Make tool
+#. Run ``make conda-environment``, if that fails you can try doing it manually
+
+    #. Change your current directory to MESMER's root directory (i.e. the one which contains ``README.rst``), ``cd mesmer``
+    #. Create a conda environment to use with MESMER ``conda create --name mesmer``
+    #. Activate your conda environment ``conda activate mesmer``
+    #. Install the development dependencies (very important, make sure your conda environment is active before doing this) ``conda install -y --file environment.yml && pip install --upgrade pip wheel && pip install -e .[dev]``
```
Comment (mathause) [inline docs/source/development.rst:44]:
```suggestion
    #. Install the development dependencies (very important, make sure your conda environment is active before doing this) ``conda install -y --file environment.yml && pip install -e .[dev]``
```

I don't think anyone will be releasing except for us, so no need for wheel (AFAIK pip is installed in conda environments per default).

File: docs/source/development.rst [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    - you can use `regex101.com <regex101.com>`_ to help write and check regular expressions, make sure the language is set to Python to make your life easy!
+
+Testing philosophy
+------------------
+
+When writing tests, we try to put them in one of two categories: integration and regression.
+Integration tests run bits of the code and assert the correct behaviour was achived.
+Some of the integration tests might run fairly big bits of code, others will be more targeted.
+Try to keep integration test files targeted and fairly small.
+We can always create fixtures to aid code reuse.
```
Comment (mathause) [inline docs/source/development.rst:107]:
```suggestion
We can always create `fixtures <https://docs.pytest.org/en/stable/explanation/fixtures.html>`__ to aid code reuse.
```

**Review by mathause — APPROVED** [review]

File: docs/source/development.rst (hunk 10 above)
Comment (znicholls) [reply docs/source/development.rst:44]: True we should be able to release using the actions

**Comment by mathause** [comment]
thanks @znicholls

**Pull request merged**